### PR TITLE
Modularize frame.rs into frames module with each frame in an individual file.

### PR DESCRIPTION
**Pull request by waieez**
Hello,

As previously discussed, we are interested in implementing more frames among other things.
In order to do this without making frame.rs an even bigger file, we should break down frame.rs into their individual files.

This PR achieves this without modifying anything else in the project.

BEFORE:
solicitmaster

AFTER:
Tests are moved into their associated frames and are still passing.
solicitframe

### CODE REVIEW

**Comment by mlalic** [comment]
Hey, this is definitely a great start toward adding more frame implementations. Thanks.

File: src/http/frame/mod.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
+/// A helper macro that unpacks a sequence of 4 bytes found in the buffer with
```
Comment (mlalic) [inline src/http/frame/mod.rs]: The module seems to have lost the doc string it had before.

File: src/http/frame/mod.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    ($buf:ident, $offset:expr, $tip:ty) => (
+        (($buf[$offset + 0] as $tip) << 24) |
+        (($buf[$offset + 1] as $tip) << 16) |
+        (($buf[$offset + 2] as $tip) <<  8) |
+        (($buf[$offset + 3] as $tip) <<  0)
+    );
+}
+
+pub use self::frames::*;
+//use self::testconfig::*;
```
Comment (mlalic) [inline src/http/frame/mod.rs]: Code that got commented out shouldn't generally end up getting merged :)

File: src/http/frame/mod.rs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+pub use self::dataframe::*;
+pub use self::settingsframe::*;
+pub use self::headersframe::*;
+
+pub mod frames;
+mod testconfig;
```
Comment (mlalic) [inline src/http/frame/mod.rs]: This module should be conditionally compiled only in a test build.

Additionally, I'd just call it `tests`, to be honest -- it's not really configuring anything, but providing the shared utilities for all the other tests.

File: src/http/frame/mod.rs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+        (($buf[$offset + 2] as $tip) <<  8) |
+        (($buf[$offset + 3] as $tip) <<  0)
+    );
+}
+
+pub use self::frames::*;
+//use self::testconfig::*;
+pub use self::dataframe::*;
+pub use self::settingsframe::*;
+pub use self::headersframe::*;
```
Comment (mlalic) [inline src/http/frame/mod.rs]: I'm not such a fan of wildcard imports in general, but especially when it comes to rexports, I'd explicitly state what it is that you intend to {ex,im}port.

In this case, it's only the frame struct that needs rexporting, so just explicitly state that.

File: src/http/frame/headersframe.rs [hunk 5]
```diff
@@ -0,0 +1,621 @@
+//! The module contains the implementation of HTTP/2 frames.
```
Comment (mlalic) [inline src/http/frame/headersframe.rs]: It seems this is where the original module-level doc string ended up -- this one should be in `frame/mod.rs` :)

File: src/http/frame/testconfig.rs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+/// followed by `pad_len` zero bytes.
+pub fn build_padded_frame_payload(data: &[u8], pad_len: u8) -> Vec<u8> {
+    let sz = 1 + data.len() + pad_len as usize;
+    let mut payload: Vec<u8> = Vec::with_capacity(sz);
+    payload.push(pad_len);
+    payload.extend(data.to_vec().into_iter());
+    for _ in 0..pad_len { payload.push(0); }
+
+    payload
+}
```
Comment (mlalic) [inline src/http/frame/testconfig.rs]: You should configure your editor to properly end files with a new line character (`\n`).

File: src/http/frame/mod.rs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+macro_rules! unpack_octets_4 {
+    ($buf:ident, $offset:expr, $tip:ty) => (
+        (($buf[$offset + 0] as $tip) << 24) |
+        (($buf[$offset + 1] as $tip) << 16) |
+        (($buf[$offset + 2] as $tip) <<  8) |
+        (($buf[$offset + 3] as $tip) <<  0)
+    );
+}
+
+pub use self::frames::*;
```
Comment (mlalic) [inline src/http/frame/mod.rs]: Come to think of it, I think there's no reason that the `frame/frames.rs` file shouldn't be fully in `frame/mod.rs`, instead.

**Comment by mlalic** [comment]
One general comment -- you should try to have the unit tests pass after every commit when possible. This has quite a few advantages, such as being able to run `git-bisect` later on without getting false positives.

In this case, after the 44c89a0 commit the unit tests end up failing. You can easily fix it here (and integrate one of the comments I left above) by simply moving `frame.rs` into `frame/mod.rs` -- two birds with one stone, you're ready to factor out the frames into their separate files (what the commit originally seems to be aiming to do) and we're not introducing the `frames.rs` file which seems to serve little purpose.

**Comment by waieez** [comment]
Ah, I'll work on fixing those asap. The only thing I'm not too sure how to do is in your last comment. Am I supposed to go back to that commit and amend the commit with those changes?

**Comment by waieez** [comment]
```
I'm getting the following error when I try to move frames.rs into mod.rs
Not quite sure what I'm doing wrong with the imports.

For now I'm explicitly importing from frames instead of using the wildcard.

$ cargo build
   Compiling solicit v0.0.1 (file:///E:/Manual%20Backup/My%20Documents/hr/thesis/solicit)
src\http\connection.rs:31:5: 31:10 error: unresolved import (maybe you meant `Frame::*`?)
src\http\connection.rs:31     Frame,
                              ^~~~~
src\http\frame/mod.rs:26:9: 26:28 error: unresolved import
src\http\frame/mod.rs:26 pub use self::dataframe::*;
                                 ^~~~~~~~~~~~~~~~~~~
src\http\frame\headersframe.rs:3:5: 3:10 error: unresolved import (maybe you meant `Frame::*`?)
src\http\frame\headersframe.rs:3     Frame,
                                     ^~~~~
src\http\frame\settingsframe.rs:3:5: 3:10 error: unresolved import (maybe you meant `Frame::*`?)
src\http\frame\settingsframe.rs:3     Frame,
                                      ^~~~~
src\http\frame\dataframe.rs:3:5: 3:10 error: unresolved import (maybe you meant `Frame::*`?)
src\http\frame\dataframe.rs:3     Frame,
                                  ^~~~~
src\http\frame\test.rs:1:13: 1:21 error: unresolved import (maybe you meant `RawFrame::*`?)
src\http\frame\test.rs:1 use super::{RawFrame};
                                     ^~~~~~~~
error: aborting due to 6 previous errors
Could not compile `solicit`.

To learn more, run the command again with --verbose.

```

**Comment by mlalic** [comment]
@waieez I'm not sure what could be causing those kinds of errors... I just tried doing myself what I suggested (simply moving `frame.rs` to `frame/mod.rs` and there's no issue with that).

One thing I notice is that your output says `frame/mod.rs` with a **forward slash** only for the `mod.rs` file -- is it possible that somehow your filesystem ended up interpreting that as the full file name, instead of a directory path?

**Comment by mlalic** [comment]
One other thing -- and I'm sorry if it's piling onto you guys -- but the ping frame feature really shouldn't be part of this pull request. For your info, any extra commits that you add to the branch that you opened the PR with are automatically included, so if you want to continue working on a new feature that depends on some of the code in an outstanding PR, you should create a new branch for it.

Luckily you can easily rebase the ping frame commits away from this branch in this case :)

This also ties into what @waieez asked about rewriting a previous commit -- yes, indeed, one of the major advantages of using git is being able to go back and amend your commits so that they end up being nice self-contained changesets, rather than the history of the repository having commits like: `Implemented feature X` and then `Added some more tests`, you can easily go back once you feel you're done and rebase the tests onto the original commit.

I could give you more info on exactly how it all works if you haven't had any experience with it just yet, but you can start off by searching around the web for `git rebase` -- and don't worry if you don't understand it all from the get go, it probably takes some time to get used to how it works, but once you do, you'll come to understand how it's a really powerful concept.

**Comment by waieez** [comment]
Ah sorry. I'll try to figure that out in a little bit. This is all a WIP and I just wanted to let you know how we had planned to proceed. Perhaps in a week or two we'll make another PR with the commits in order. In the meantime, we're going to try our best to manage the commit history going forwards.

**Pull request closed without merging**